homeassistant/components/search/__init__.py

```diff
@@ -0,0 +1,180 @@
+"""The Search integration."""
+from collections import defaultdict
+
+import voluptuous as vol
+
+from homeassistant.components import websocket_api
+from homeassistant.components.homeassistant import scene
+from homeassistant.core import HomeAssistant, split_entity_id
+from homeassistant.helpers import device_registry, entity_registry
+
+DOMAIN = "search"
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Search component."""
+    websocket_api.async_register_command(hass, websocket_search)
+    return True
+
+
+@websocket_api.async_response
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "search",
+        vol.Required("item_type"): vol.In(
+            ("area", "device", "entity", "script", "scene", "automation")
+        ),
+        vol.Required("item_id"): str,
+    }
+)
+async def websocket_search(hass, connection, msg):
+    """Handle search."""
+    searcher = Searcher(
+        hass,
+        await device_registry.async_get_registry(hass),
+        await entity_registry.async_get_registry(hass),
+    )
+    connection.send_result(
+        msg["id"], await searcher.search(msg["item_type"], msg["item_id"])
+    )
+
+
+class Searcher:
+    """Find related things.
+
+    Few rules:
+    Scenes, scripts, automations and config entries will only be expanded if they are
+    the entry point. They won't be expanded if we process them. This is because they
+    turn the results into garbage.
+    """
+
+    # These types won't be further explored
+    DONT_RESOLVE = {"scene", "automation", "script", "config_entry"}
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        device_reg: device_registry.DeviceRegistry,
+        entity_reg: entity_registry.EntityRegistry,
+    ):
+        """Search results."""
+        self.hass = hass
+        self._device_reg = device_reg
+        self._entity_reg = entity_reg
+        self.results = defaultdict(set)
+        self._to_resolve = set()
+
+    async def search(self, item_type, item_id):
```

Comment:
I believe this should be
```suggestion
    @callback
    def async_search(self, item_type, item_id):
```